proxy/src/telemetry/metrics/latency.rs

```diff
@@ -0,0 +1,291 @@
+        for ref mut bucket in &mut self.buckets[..] {
+            if bucket.add(t) { return true; }
+        }
+        false
+    }
+}
+
+impl<T> Buckets<T, LeafBucket<T>> 
+where
+    T: PartialOrd
+{
+    /// Construct a set of five linear [`LeafBucket`]s with width `width`.
+    /// 
+    /// [`LeafBucket`]: struct.LeafBucket.html
+    fn linear(width: u32) -> Self
+    where T: From<u32> {
+        Buckets {
+            max: T::from(width * 5),
+            buckets: [
+                // TODO: construct this from a range argument?
+                LeafBucket::new(width),
+                LeafBucket::new(width * 2),
+                LeafBucket::new(width * 3),
+                LeafBucket::new(width * 4),
+                LeafBucket::new(width * 5),
+            ]
+        }
+
+    }
+}
+
+impl<'a, T> IntoIterator for &'a Buckets<T>
+where
+    T: PartialOrd,
+{
+    type Item = &'a LeafBucket<T>;
+    type IntoIter = slice::Iter<'a, LeafBucket<T>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.buckets.into_iter()
+    }
+
+}
+
+impl<'a, T> IntoIterator for &'a Buckets<T, Buckets<T>>
+where
+    T: PartialOrd,
+{
+    type Item = &'a LeafBucket<T>;
+    // TODO: remove box
+    type IntoIter = Box<Iterator<Item=Self::Item> + 'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Box::new(
+            self.buckets
+                .into_iter()
+                .flat_map(|bucket| bucket.into_iter())
+        )
+
+    }
+
+}
+
+// ===== impl Latency =====
+
+const SEC_TO_MS: u32 = 1_000;
+const SEC_TO_TENTHS_OF_A_MS: u32 = SEC_TO_MS * 10;
+const TENTHS_OF_MS_TO_NS: u32 =  MS_TO_NS / 10;
+/// Conversion ratio from milliseconds to nanoseconds.
+pub const MS_TO_NS: u32 = 1_000_000;
+
+impl From<Duration> for Latency {
+    fn from(dur: Duration) -> Self {
+        let secs = dur.as_secs();
+        // checked conversion from u64 -> u32.
+        let secs =
+            if secs >= u64::from(u32::MAX) {
+                None
+            } else {
+                Some(secs as u32)
+            };
+        // represent the duration as tenths of a ms.
+        let tenths_of_ms =
+            secs.and_then(|as_secs|
+                // convert the number of seconds to tenths of a ms, or
+                // None on overflow.
+                as_secs.checked_mul(SEC_TO_TENTHS_OF_A_MS)
+            )
+                .and_then(|as_tenths_ms| {
+                    // convert the subsecond part of the duration (in ns) to
+                    // tenths of a millisecond.
+                    let subsec_tenths_ms = dur.subsec_nanos() / TENTHS_OF_MS_TO_NS;
+                    as_tenths_ms.checked_add(subsec_tenths_ms)
+                })
+                .unwrap_or_else(|| {
+                    debug!("{:?} too large to represent as tenths of a millisecond!", dur);
+                    u32::MAX
+                });
+        Latency(tenths_of_ms)
+    }
+}
+
+impl From<u32> for Latency {
+    #[inline] 
```

Comment:
i'm generally very wary of telling the compiler how to do its job. What's the argument for including these?

Reply:
honestly, mainly because it reminds _me_ they'll probably be inlined & makes a function call seem less scary...AFAIK, `#[inline]` (vs. `#[inline(always)]`) is a fairly gentle hint? i can remove them.

Reply:
I'm not particularly opposed to it. I just want to have a good rule in my mind about when they should be used and I'm not really clear on that. Without such a rule, I guess I'd prefer that they be removed so I don't have to wonder why some methods were marked as inline versus others -- i.e. i'm sure we have many other small methods without the annotation.  So whatever we're going to do, consistency is king.

Reply:
that's totally reasonable! I've removed them.